### PR TITLE
Add a design docs page

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by roidelapluie** [comment]
I will add more design docs once we agree on the principle.

**Comment by RichiH** [comment]
I think the thing is good, just those nits.

Sent by mobile; please excuse my brevity.

**Comment by beorn7** [comment]
Looks great in general.

Do you want to fill in more design docs here right now? Or do that in later PRs?

**Comment by beorn7** [comment]
In case it's useful, here my brainstorm:
* prometheus/client_java & micrometer (status: under review, I guess): https://docs.google.com/document/d/1vROky2aIw3kAllfi95gwDJy5P2DyWnCihsjPXGpLwwo/edit# 
* Configuration handling in exporters and Prometheus 3.x (status: under review): https://docs.google.com/document/d/1BK_Gc3ixoWyxr9F5qGC07HEcfDPtb6z96mfqoGyz52Y/edit#heading=h.zcana55kzmxi
* http_sd_configs (status: under review, being implemented): https://docs.google.com/document/d/1tVeuzjpU4-TiYPNWJXKmcyIuZF6A2tUq270RbBT5zho/edit
* Prom Agent (status: approved, more or less, and being implemented): https://docs.google.com/document/d/1cCcoFgjDFwU2n823tKuMvrIhzHty4UDyn0IcfUHiyyI/edit
* Sparse high-resolution histograms (status: approved): https://docs.google.com/document/d/1cLNv3aufPZb3fNfaJgdaRBZsInZKKIHo9E6HinJVbpM/edit
* OpenMetrics transition (status: TODO): (as per dev-summit consensus: "We will create a design document exploring upgrade paths.")
* Metadata (status: empty, oops): https://docs.google.com/document/d/1XiZePSjwU4X5iaIgCIvLzljJzl8lRAdvputuborUcaQ/edit
* topk/bottomk aggregation over time (status: implemented): https://docs.google.com/document/d/1uSbD3T2beM-iX4-Hp7V074bzBRiRNlqUdcWP6JTDQSs/edit
* Semantics of muting in Alertmanager (status: TODO): (as per dev-summit consensus: "We will create a document for the community to collect current pain points and desired behaviour"
* First-class network monitoring support in the Prometheus & Grafana ecosystem (status: under review): https://docs.google.com/document/d/1oEpjiWfTHF352NCAOGolwij3EIkrprCkdQmaQMpjg4M/edit
* Persist Retroactive Rules (status: approved, partially implemented): https://docs.google.com/document/d/16s_-RxYwQYcb4G4mvpmjPolEbD24o54a1LPqJ538Vhc/edit
* TSDB Head Improvements (status: implemented, at least partially, or all? paging @codesome ): https://docs.google.com/document/d/184urkLQnM7rqLmGvS66I15jU2pyPk_qwd8v_qDlXczo/edit# and https://docs.google.com/document/d/1pnEsxB0CDLOxQipGw_vhkJpoDZfZPs_KjqCriumDAXQ/edit

(This is not meant to be a complete list. ;)

**Comment by beorn7** [comment]
* Secure Alertmanager cluster traffic (status: being implemented in https://github.com/prometheus/alertmanager/pull/2237): https://github.com/prometheus/alertmanager/blob/master/doc/design/secure-cluster-traffic.md

**Comment by codesome** [comment]
> TSDB Head Improvements (status: implemented, at least partially, or all? paging @codesome ): https://docs.google.com/document/d/184urkLQnM7rqLmGvS66I15jU2pyPk_qwd8v_qDlXczo/edit# and https://docs.google.com/document/d/1pnEsxB0CDLOxQipGw_vhkJpoDZfZPs_KjqCriumDAXQ/edit


Only the m-mapping of "full" chunks is done among them. The idea about "partial chunks" is in-progress (the [snapshot PR](https://github.com/prometheus/prometheus/pull/7229)).

Plus, everything in that doc is not planned to be implemented (for example the number 5 in second doc about series and tombstones). And the doc might/will not be 100% in sync with what was finally implemented or being implemented.

**Comment by beorn7** [comment]
Thanks @codesome . 

I guess a certain degree of "we did this differently after all" is inherent to most design docs. So I don't think we have to call that out explicitly (but perhaps it's worth mentioning in the preamble on this page?).

File: content/docs/introduction/design-doc.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+When contributing large changes or features to the Prometheus ecosystem, design
+documents are written by individuals or groups. Those documents are proposed for
+review and approval to the community.
+
+This page is a list of the design documents we know of. If you create a new
+design document, or some documents are not listed, please open a [pull
+request](https://github.com/prometheus/docs/) to add it to this list.
+
+Design doc dot not always reflect exactly what has been implemented, and
```
Comment (beorn7) [inline content/docs/introduction/design-doc.md:16]: Design doc dot not… → Design docs do not…

File: content/docs/introduction/design-doc.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+If you want to pick a design doc in `TODO` or implement a proposed design doc,
+please reach out to the [developers mailing list](/community) first to make sure
+no one else has picked that task already, and the design doc is approved and
+still relevant.
+
+<table>
+  <tr><th>Document</th><th>Initial date</th><th>Status</th><th>Pull requests</th></tr>
+  <tr><td><a href="https://github.com/prometheus/alertmanager/blob/master/doc/design/secure-cluster-traffic.md">Secure Alertmanager cluster traffic</a></td><td>2019-02-21</td><td>Approved</td><td><a href="https://github.com/prometheus/alertmanager/pull/2237">alertmanager#2237</a></td></tr>
+  <tr><td>TSDB Head Improvements (<a href="document 1">https://docs.google.com/document/d/184urkLQnM7rqLmGvS66I15jU2pyPk_qwd8v_qDlXczo/edit#heading=h.ow34omg0frla</a> - <a href="document 2">https://docs.google.com/document/d/1pnEsxB0CDLOxQipGw_vhkJpoDZfZPs_KjqCriumDAXQ/edit#heading=h.qsf5sgqoqemc</a>)</td><td>2019-12-09</td><td>Partially implemented</td><td></td></tr>
```
Comment (beorn7) [inline content/docs/introduction/design-doc.md:28]: Link text and URL are switched here. (URL needs to be the `href` and "document 1/2"  needs to be the text content of the tag.)

File: content/docs/introduction/design-doc.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  <tr><td><a href="https://docs.google.com/document/d/1uSbD3T2beM-iX4-Hp7V074bzBRiRNlqUdcWP6JTDQSs/edit">topk/bottomk aggregation over time</a></td><td>2020-09-30</td><td>Implemented</td><td><a href="https://github.com/prometheus/prometheus/pull/8121">#8121</a> <a href="https://github.com/prometheus/prometheus/pull/8425">#8425</a></td></tr>
+  <tr><td><a href="https://docs.google.com/document/d/1tVeuzjpU4-TiYPNWJXKmcyIuZF6A2tUq270RbBT5zho/edit">http_sd_configs</a></td><td>2021-02-26</td><td>Under review</td><td><a href="https://github.com/prometheus/prometheus/pull/8839">#8839</a></td></tr>
+  <tr><td><a href="https://docs.google.com/document/d/1vROky2aIw3kAllfi95gwDJy5P2DyWnCihsjPXGpLwwo/edit#">prometheus/client_java & micrometer</a></td><td>2021-02-26</td><td>Under review</td><td></td></tr>
+  <tr><td><a href="https://docs.google.com/document/d/1oEpjiWfTHF352NCAOGolwij3EIkrprCkdQmaQMpjg4M/edit"></a>First-class network monitoring support in the Prometheus & Grafana ecosystem</td><td>2021-02-25</td><td>Under review</td><td></td></tr>
```
Comment (beorn7) [inline content/docs/introduction/design-doc.md:33]: The `</a>` has to be moved after the title.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks for adding all those links to the table.

I found a few syntax issues, see comments.

In different news: Are you aware that we can use markdown tables on the site? See https://github.com/prometheus/docs/blob/master/content/docs/practices/histograms.md .

**Comment by beorn7** [comment]
An now there is the [goreleaser design doc](https://docs.google.com/document/d/16LOT2wK-jntlU-EFADfaEF3YbKH81U9Zl_PvSu4qVwo/edit#heading=h.bupciudrwmna).

List is growing quickly. I guess it makes sense to merge this rather sooner than later so that we all can iterate on it easily.

(But see my syntax remarks above. And my question if we can't just use markdown tables, which would avoid those syntax quirks altogether.)

**Comment by beorn7** [comment]
I guess we should also add a TODO entry for reconsidering extrapolation with range selectors (rate/increase/delta) AKA "the (x)rate controversy". At the [dev summit 2020/2](https://docs.google.com/document/d/1vhXKpCNY0k2cbm0g10uM2msXoMoH8CTwrg_dyqsFUKo/edit), we stated: "We want to rethink this from first principles. We will collect our research, thoughts, considerations, and results in 'Prometheus is not feature complete' and discuss at a future dev summit."

I think the topic deserves a fully fledged design doc. I guess we just hadn't developed the habit back then and thought, we could shove it in that existing "Prometheus is not feature complete" document.

**Review by beorn7 — APPROVED** [review]
Looks great.

I'd say, let's merge and iron out any other nits incrementally later.

File: content/docs/introduction/design-doc.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+When contributing large changes or features to the Prometheus ecosystem, design
+documents are written by individuals or groups. Those documents are proposed for
+review and approval to the community.
+
+This page is a list of the design documents we know of. If you create a new
+design document, or some documents are not listed, please open a [pull
+request](https://github.com/prometheus/docs/) to add it to this list.
+
+Design doc do not always reflect exactly what has been implemented, and
```
Comment (RichiH) [inline content/docs/introduction/design-doc.md:16]:
```suggestion
Design docs do not always reflect exactly what has been implemented, and
```

File: content/docs/introduction/design-doc.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+implementation details might have changed since a feature was merged. Design
+docs are not considered documentation.
```
Comment (RichiH) [inline content/docs/introduction/design-doc.md:18]:
```suggestion
docs are not considered documentation and can not define a standard.
```

**Review by RichiH — CHANGES_REQUESTED** [review]
Minor nits; looks good otherwise, thanks!

**Review by RichiH — APPROVED** [review]

**Comment by RichiH** [comment]
Thanks!

**Comment by roidelapluie** [comment]
> Thanks!

I hated this :)

**Comment by roidelapluie** [comment]
Note: in my last push, I have used "unbreakable dashes" instead of "dashes" for the dated. It keeps them on the same line in the page, I think it renders better.

**Pull request merged**